File: baekjoon/python/1260.py
from collections import deque
# import sys
# sys.stdin = open("1260_input.txt", 'r')


n, m, v = map(int, input().split())
graph = [[] for _ in range(n+1)]
visited = [0] * (n+1)
for _ in range(m):
    start, end = map(int, input().split())
    graph[start].append(end)
    graph[end].append(start)
for i in range(1, n+1):
    graph[i].sort()

# print(graph)


def init():
    global visited
    visited = [0] * (n+1)
    print()


def dfs(start):
    s = []
    s.append(start)
    # visited[start] = 1
    # print(start, end=" ")
    while s:
        # 현재지점
        vertex = s.pop()
        # 방문 했다면
        if visited[vertex]:
            continue
        # 방문하지 않았다면
        if not visited[vertex]:
            visited[vertex] = 1
            print(vertex, end=" ")
            # 주변 노드 방문
            for nv in graph[vertex][::-1]:
                s.append(nv)


def dfs_rcs(start):
    # 방문한 곳이면
    if visited[start]:
        return False
    # 방문 안했으면
    if not visited[start]:
        # 방문처리
        visited[start] = 1
        print(start, end=" ")
        # 주변 노드 방문
        for node in graph[start]:
            dfs_rcs(node)
    return True


def bfs(start):
    q = deque()
    q.append(start)
    visited[start] = 1
    print(start, end=" ")
    while q:
        # 현재지점
        vertex = q.popleft()
        # 주변 노드 방문
        for nv in graph[vertex]:
            # 방문 했다면
            if visited[nv]:
                continue
            # 방문하지 않았다면
            if not visited[nv]:
                visited[nv] = 1
                q.append(nv)
                print(nv, end=" ")


# dfs_rcs(v)
# init()
dfs(v)
init()
bfs(v)
init()
# print(graph, visited)


# print(bfs(v))

File: baekjoon/cpp/main_acc/1260/1260.cpp
#include <bits/stdc++.h>
using namespace std;
vector<int> graph[1001];
bool visited[1001] = {
    false,
};

vector<int> res_bfs;
vector<int> res_dfs;
void bfs(int node)
{
    queue<int> q;
    q.push(node);
    visited[node] = true;
    while (!q.empty())
    {
        int out = q.front();
        q.pop();
        res_bfs.push_back(out);
        for (const auto &elem : graph[out])
        {
            if (!visited[elem])
            {
                visited[elem] = true;
                q.push(elem);
            }
        }
    }
}
void dfs(int node)
{
    visited[node] = true;
    res_dfs.push_back(node);
    for (const auto &elem : graph[node])
    {
        if (!visited[elem])
        {
            dfs(elem);
        }
    }
}
// int dfs_rcs(int node)
// {
//     if (visited[node])
//         return false;
//     else
//     {
//         visited[node] = true;
//         cout << node << " ";
//         for (int &elem : graph[node])
//         {
//             dfs_rcs(elem);
//         }
//         return true;
//     }
// }
// void dfs_stack(int node)
// {
//     vector<int> s;
//     s.push_back(node);
//     // visited[node] = true;
//     // cout << node << " ";
//     while (!s.empty())
//     {
//         int out = s.back();
//         s.pop_back();
//         if (!visited[out])
//         {
//             visited[out] = true;
//             cout << out << " ";
//             for (int i = graph[out].size() - 1; i != -1; --i)
//             {
//                 int &elem = graph[out][i];
//                 s.push_back(elem);
//             }
//         }
//     }
//     cout << endl;
// }

int main()
{
    std::ios_base::sync_with_stdio(false);
    freopen("1260_input.txt", "r", stdin);

    int n, m, v;
    cin >> n >> m >> v;
    // cout << n << m << v << endl;
    for (int i = 0; i < m; i++)
    {
        int a, b;
        cin >> a >> b;
        graph[a].push_back(b);
        graph[b].push_back(a);
    }

    for (int i = 1; i <= n; i++)
    {
        sort(graph[i].begin(), graph[i].end());
    }

    // for (int i = 1; i <= n; i++)
    // {
    //     cout << i << " : ";
    //     for (auto &elem : graph[i])
    //     {
    //         cout << elem << " ";
    //     }
    //     cout << endl;
    //     // cout << graph[i].size() << endl;
    // }
    // dfs_stack(v);
    // memset(visited, false, sizeof(visited));
    // dfs_rcs(v);
    // cout << endl;
    dfs(v);
    memset(visited, false, sizeof(visited));
    bfs(v);
    for (const auto &elem : res_dfs)
    {
        cout << elem << " ";
    }
    cout << endl;
    for (const auto &elem : res_bfs)
    {
        cout << elem << " ";
    }
    cout << endl;
    // cout << visited[0] << endl;

    // cout << visited[0] << endl;
    return 0;
}
